File: counting-bits/optimized.java
public class Solution {
    /**
     * Returns an array of length n + 1 where each index i represents the number of 1's in the binary representation of i.
     * 
     * @param n The input integer.
     * @return An array of length n + 1 with the number of 1's in the binary representation of each index.
     */
    public int[] countBits(int n) {
        // Initialize the result array with a size of n + 1.
        int[] result = new int[n + 1];
        
        // Base case: The number of 1's in the binary representation of 0 is 0.
        result[0] = 0;
        
        // Iterate over the range from 1 to n (inclusive).
        for (int i = 1; i <= n; i++) {
            // Calculate the number of 1's in the binary representation of i using the formula:
            // result[i] = result[i >> 1] + (i & 1)
            // This formula works because the number of 1's in the binary representation of i is equal to the number of 1's in the binary representation of i >> 1 (right shift by 1) plus the least significant bit of i (i & 1).
            result[i] = result[i >> 1] + (i & 1);
        }
        
        // Return the result array.
        return result;
    }
}

File: number-of-1-bits/optimized.c
/**
 * Returns the number of set bits in the binary representation of a given positive integer.
 *
 * @param n A positive integer.
 * @return The number of set bits in the binary representation of n.
 */
int hammingWeight(int n) {
    int count = 0;
    // Continue until all bits are processed
    while (n) {
        // Clear the least significant set bit
        n &= (n - 1);
        // Increment the count of set bits
        count++;
    }
    return count;
}